contributors/design-proposals/controller-ref.md

```diff
@@ -1,101 +1,375 @@
+behavior, based on OwnerReferences.
+
+In addition, if a controller finds that it owns an object that no longer matches
+its selector, it should orphan the object by removing itself from the object's
+OwnerReferences list. Since ControllerRef is just a special type of
+OwnerReference, this also means the ControllerRef is removed.
+
+## Watches
+
+Many controllers use watches to *sync* each controller instance (prompting it to
+reconcile desired and actual state) as soon as a relevant event occurs for one
+of its controlled objects, as well as to let controllers wait for asynchronous
+operations to complete on those objects.
+The controller subscribes to a stream of events about controlled objects
+and routes each event to a particular controller instance.
+
+Previously, the controller used only label selectors to decide which
+controller to route an event to. If multiple controllers had overlapping
+selectors, events might be misrouted, causing the wrong controllers to sync.
+Controllers could also freeze because they keep waiting for an event that
+already came but was misrouted, manifesting as `kubectl` commands that hang.
+
+Some controllers introduced a workaround to break ties. For example, they would
+sort all controller instances with matching selectors, first by creation
+timestamp and then by name, and always route the event to the first controller
+in this list. However, that did not prevent misrouting if the overlapping
+controllers were of different types. It also only worked while controllers
+themselves assigned ownership over objects using the same tie-break rules.
+
+Now that controller ownership is defined in terms of ControllerRef,
+controllers should use the following guidelines for responding to watch events:
+
+* If the object has a ControllerRef:
+  * Sync only the referenced controller.
+  * Update `expectations` counters for the referenced controller.
+  * If an *Update* event removes the ControllerRef, sync any controllers whose
+    selectors match to give each one a chance to adopt the object.
+* If the object is an orphan:
+  * *Add* event
+    * Sync any controllers whose selectors match to give each one a chance to
+      adopt the object.
+    * Do *not* update counters on `expectations`.
+      Controllers should never be waiting for creation of an orphan because
+      anything they create should have a ControllerRef.
+  * *Delete* event
+    * Do *not* sync any controllers.
+      Controllers should never care about orphans disappearing.
+    * Do *not* update counters on `expectations`.
+      Controllers should never be waiting for deletion of an orphan because they
+      are not allowed to delete objects they don't own.
+  * *Update* event
+    * If labels changed, sync any controllers whose selectors match to give each
+      one a chance to adopt the object.
+
+## Default garbage collection policy
+
+Most controllers that implement ControllerRef should have a
```

Comment:
I'm not sure that's true for new controllers. The reason why this default was needed was backward compatibility. If not for that "cascade by default" sounds like a more sane default for me.